processor/tailsamplingprocessor/sampling/logical_or_attribute_evaluator.go

```diff
@@ -0,0 +1,251 @@
+	// MinValue is the minimum value of the attribute to be considered a match.
+	MinValue int64 `mapstructure:"min_value"`
+	// MaxValue is the maximum value of the attribute to be considered a match.
+	MaxValue int64 `mapstructure:"max_value"`
+}
+
+// StringAttributeEvaluatorCfg holds the configurable settings to create a string attribute evaluator
+type StringAttributeEvaluatorCfg struct {
+	// Tag that the filter is going to be matching against.
+	Key string `mapstructure:"key"`
+	// Values is the set of attribute values that if any is equal to the actual attribute value to be considered a match.
+	Values []string `mapstructure:"values"`
+	// Specifies the final sampling decision on match. Defaults to sampled.
+	// Valid values are: sampled, notsampledfinal
+	DecisionOnMatch string `mapstructure:"decision_on_match"`
+}
+
+type logicalOrAttributeEvaluator struct {
+	logger             *zap.Logger
+	keyToEvaluatorsMap map[string][]attributeEvaluator
+}
+
+type stringAttributeEvaluator struct {
+	key             string
+	values          map[string]struct{}
+	decisionOnMatch Decision
+	logger          *zap.Logger
+}
+
+type numericAttributeEvaluator struct {
+	key                string
+	minValue, maxValue int64
+	logger             *zap.Logger
+}
+
+type attributeEvaluator interface {
+	isMatch(value pdata.AttributeValue) (Decision, error)
+	attributeKey() string
+}
+
+func newLogicalOrAttributeEvaluator(logger *zap.Logger, evaluators []attributeEvaluator) PolicyEvaluator {
+	// Build up a map where the key is the attribute name and value is an array of evaluators for that attribute value.
+	// For each attribute iterated over, the key is used to locate evaluators.
+	// Worst case, there will be n lookups in this map followed by 1 or more evaluations against the attribute value.
+	keyToEvaluatorsMap := make(map[string][]attributeEvaluator)
+	for _, evaluator := range evaluators {
+		key := evaluator.attributeKey()
+		var keySamplers []attributeEvaluator
+
+		if k, ok := keyToEvaluatorsMap[key]; !ok {
+			keySamplers = make([]attributeEvaluator, 0)
+		} else {
+			keySamplers = k
+		}
+
+		keyToEvaluatorsMap[key] = append(keySamplers, evaluator)
+	}
+
+	return &logicalOrAttributeEvaluator{
+		keyToEvaluatorsMap: keyToEvaluatorsMap,
+		logger:             logger,
+	}
+}
+
+// NewLogicalOrAttributeEvaluator creates a policy evaluator that samples traces based on one or more attribute evaluators OR'd together
+func NewLogicalOrAttributeEvaluator(logger *zap.Logger, evaluatorsConfig []SingleAttributeEvaluatorCfg) (PolicyEvaluator, error) {
+	// Build the evaluators from config
+	evaluators := make([]attributeEvaluator, len(evaluatorsConfig))
+	for i, evaluatorCfg := range evaluatorsConfig {
+		var evaluator attributeEvaluator = nil
+		if evaluatorCfg.StringAttributeCfg.Key != "" {
+			stringEvaluatorCfg := evaluatorCfg.StringAttributeCfg
+
+			decisionOnMatch := Sampled
+			switch stringEvaluatorCfg.DecisionOnMatch {
+			case "":
+			case "sampled":
```

Comment:
man this had me so confused coming from c#. I get why you did this (so that you can do error handling), but wondering if there is a way to make this more readable to c# heavy folks.